Lib/calendar.py

```diff
@@ -548,13 +548,25 @@ def formatyearpage(self, theyear, width=3, css='calendar.css', encoding=None):
 class different_locale:
     def __init__(self, locale):
         self.locale = locale
+        self.oldlocale = None
 
     def __enter__(self):
-        self.oldlocale = _locale.getlocale(_locale.LC_TIME)
-        _locale.setlocale(_locale.LC_TIME, self.locale)
+        self.oldlocale = _locale.setlocale(_locale.LC_TIME, None)
+        if self.locale is not None:
+            _locale.setlocale(_locale.LC_TIME, self.locale)
+        else:
+            if self.oldlocale in ("C", "POSIX"):
+                # The locale does not seem to be configured:
+                # get the user preferred locale.
+                _locale.setlocale(_locale.LC_TIME, "")
+            else:
+                # Otherwise simply use the current LC_TIME locale.
+                # The locale is not changed, __exit__() does nothing.
+                self.oldlocale = None
```

Comment:
The problem with this is that it can change based on environment variables:

```python
>>> locale.setlocale(locale.LC_TIME, '')
'ru_RU.utf8'
>>> os.environ['LC_ALL'] = 'en_US.utf8'
>>> locale.setlocale(locale.LC_TIME, '')
'en_US.utf8'
```
But the instance is supposed to *capture* the preferred locale when it's created.

Reply:
Why is it so important for you to get the locale when Python is started? As you shown, there are many ways to change locales. And using an locale set before all the changes is irrelevant, so it's now outdated, no?

Reply:
The previous code based on `getdefaultlocale()` was a one time check of `LC_ALL`, `LC_CTYPE`, and `LANG`. This was technically wrong, but in practice the default `LC_TIME` is usually the same as the default `LC_CTYPE`.

I think it's correct to use the current `LC_TIME` locale, if it was configured already. Not doing so was a bug. However, it should try to remain close to the previous default behavior when instantiating `LocaleTextCalendar()` when `LC_TIME` is the initial "C" value, which is usually the case if a script doesn't explicitly set `LC_ALL` or `LC_TIME`. 

But changing it to re-evaluate the default via `setlocale(LC_TIME, "")` for every `formatweekday()` call is going too far. That depends on the current values of the environment variables `LC_ALL`, `LC_TIME`, and `LANG`, which is very different from the one-time `getdefaultlocale()` call in the previous code.